<a href="https://colab.research.google.com/github/firdaaacy/Praktikum-PPM/blob/main/Praktikum%205.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Praktikum Pengantar Pembelajaran Mesin


---
## Bab 5. Klasifikasi Decision Tree


### 1) Import Data

Praktikum kali ini menggunakan dataset [Car Evaluation Dataset](https://archive.ics.uci.edu/ml/datasets/Car+Evaluation) dari UCI Machine Learning Repository. Dataset ini telah digunakan pada praktikum sebelumnya. Detail fitur dapat Anda pelajari pada link yang tersedia. 

Unduh dataset yang akan digunakan pada praktikum kali ini. Anda dapat menggunakan aplikasi wget untuk mendowload dataset dan menyimpannya dalam Google Colab. Jalankan cell di bawah ini untuk mengunduh dataset

In [4]:
#mengunduh dataset yang akan digunakan
!wget https://dataset-ppm.s3.amazonaws.com/car_sample.csv

--2021-07-09 13:58:02--  https://dataset-ppm.s3.amazonaws.com/car_sample.csv
Resolving dataset-ppm.s3.amazonaws.com (dataset-ppm.s3.amazonaws.com)... 52.216.17.96
Connecting to dataset-ppm.s3.amazonaws.com (dataset-ppm.s3.amazonaws.com)|52.216.17.96|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 7226 (7.1K) [text/csv]
Saving to: ‘car_sample.csv.1’

car_sample.csv.1    100%[===================>]   7.06K  --.-KB/s    in 0s      

2021-07-09 13:58:02 (122 MB/s) - ‘car_sample.csv.1’ saved [7226/7226]



Setelah dataset berhasil diunduh, langkah berikutnya adalah membaca dataset dengan memanfaatkan fungsi **readcsv** dari library pandas. Lakukan pembacaan berkas csv ke dalam dataframe dengan nama **data** menggunakan fungsi **readcsv**. Jangan lupa untuk melakukan import library pandas terlebih dahulu


In [5]:
#melakukan import library yang akan digunakan
import pandas as pd #pemanggilan library pandas dengan alias pemanggilan pd
import numpy as np #pemanggilan library numpy dengan alias pemanggilan np

#melakukan pembacaan dataset yang sudah diunduh menggunakan pandas
data = pd.read_csv('car_sample.csv')



Cek isi dataset Anda dengan menggunakan perintah **head()**

In [6]:
#mencetak 5 data teratas
data.head()

buying maint lug_boot safety class
0  vhigh   low    small   high   acc
1  vhigh   low      big   high   acc
2  vhigh   med      big    med   acc
3    med   med      med   high   acc
4    low  high      med   high   acc

## 2) Membagi data menjadi data latih dan data uji

Metode pembelajaran mesin memerlukan dua jenis data :


1.   Data latih : Digunakan untuk proses training metode klasifikasi
2.   Data uji : Digunakan untuk proses evaluasi metode klasifikasi

Data uji dan data latih perlu dibuat terpisah (mutualy exclusive) agar hasil evaluasi lebih akurat.

Data uji dan data latih dapat dibuat dengan cara membagi dataset dengan rasio tertentu, misalnya 80% data latih dan 20% data uji.

Library Scikit-learn memiliki fungsi [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) pada modul **model_selection** untuk membagi dataset menjadi data latih dan data uji. Bagilah dataset anda menjadi dua, yaitu **data_latih** dan **data_uji**. Agar pengacakan data dilakukan secara konstan, parameter **random_state** diisi dengan nilai integer tertentu, pada praktikum ini diset 101. Kemudian, nilai indeks pada data latih dan data uji diatur ulang agar berurutan nilainya


In [7]:
#melakukan import sebagian library yang akan digunakan
#pada kasus  ini hanya mengimport train_test_split dari modul sklearn.model_selection
from sklearn.model_selection import train_test_split
#melakukan pembagian dataset dengan pembagian 20% data uji dan 80% data latih dengan nilai random state 101 agar pengacakan konstan
data_latih, data_uji = train_test_split(data, test_size = 0.2, random_state = 101)
#mengatur ulang index data latih
data_latih.reset_index(drop=True)

buying maint lug_boot safety  class
0      low  high    small   high    acc
1      med   low      big   high  vgood
2      low   low    small   high   good
3     high  high      med   high  unacc
4     high   low    small   high  unacc
..     ...   ...      ...    ...    ...
203    med   med      med   high  vgood
204    low   low      med    med   good
205    low   med      med    med   good
206   high  high      med   high    acc
207  vhigh   low      big   high    acc

[208 rows x 5 columns]

In [8]:
#mengatur ulang index data uji
data_uji.reset_index(drop=True)

buying  maint lug_boot safety  class
0   vhigh    med      big    med    acc
1    high   high      med   high    acc
2     med    low      big    med   good
3     low    med    small   high   good
4     med    low      big   high  vgood
5     low   high      big   high  vgood
6   vhigh    low      med   high    acc
7   vhigh    med      med   high  unacc
8     low    med      med   high  vgood
9    high    med      big    med    acc
10    low    med      med   high   good
11    med    low      big    med  unacc
12    low    med    small   high   good
13    med    low      med   high  vgood
14   high    low    small    med  unacc
15    med    low      big    med   good
16    med    low    small   high   good
17   high    low      big   high    acc
18  vhigh   high      med    low  unacc
19    low  vhigh      big    low  unacc
20    med    med      big   high  vgood
21  vhigh    low      med   high    acc
22    low    med      big   high  vgood
23    med    low      med    med   good
24    low   high      big    low  unacc
25  vhigh  vhigh      med    med  unacc
26    low  vhigh      med   high    acc
27    low   high      med   high    acc
28    med    low    small   high   good
29    low    med      big   high  vgood
30    low    med      med   high   good
31    low    med      big    med   good
32    med    low      big    med   good
33    low    med    small   high   good
34   high   high    small    med  unacc
35    med    med      big   high  vgood
36  vhigh    med    small    med  unacc
37    med    low      big    med   good
38  vhigh  vhigh      med   high  unacc
39   high    med    small   high  unacc
40    low  vhigh      med    low  unacc
41    low    low      big    med   good
42    low    med      big    med   good
43    med    med      med   high  vgood
44    med   high      med   high    acc
45    low    low      big   high  vgood
46    low    med      big    med   good
47    med    med      med    med    acc
48    low    med      big   high  vgood
49    med    med      med   high  vgood
50    low    med      big   high  vgood
51    low   high      big   high  vgood

Tampilkan banyaknya data pada **data_latih** dan **data_uji**. Seharusnya **data_latih** terdiri dari 208 data, dan **data_uji** terdiri dari 52 data

In [9]:
#menampilkan ukuran data uji
print(data_uji.shape[0])
#menampilkan ukuran data latih
print(data_latih.shape[0])

52
208


## 3) Menghitung Gini

Nilai Gini merupakan salah satu kriteria penentu variabel apa yang akan digunakan untuk membentuk cabang pada decision tree. Variabel dengan nilai Gini terbesar akan digunakan sebagai pembentukan cabang

Buatlah fungsi bernama **hitung_gini** yang berfungsi menghitung nilai Gini dari suatu nilai pada sebuah variabel

In [10]:
#menghitung nilai gini 
def hitung_gini(kolom_kelas):
  #menyimpan nilai kolom_kelas dalam variabel elemen dengan menghilangkan duplikat
  #menyimpan banyak data
  elemen, banyak = np.unique(kolom_kelas, return_counts=True)
  #mengitung nilai gini 
  nilai_gini = 1-np.sum([(banyak[i]/np.sum(banyak))**2 for i in range(len(elemen))])
  #mengembalikan nilai_gini
  return nilai_gini

Buatlah fungsi bernama **gini_split** yang digunakan untuk menghitung nilai Gini keseluruhan dari sebuah variabel.

In [11]:
#menghitung nilai gini total dari sebuah variabel
def gini_split(data, nama_fitur_split, nama_fitur_kelas):
  #menyimpan nilai kolom_kelas dalam variabel elemen dengan menghilangkan duplikat
  #menyimpan banyak data
  nilai, banyak = np.unique(data[nama_fitur_split], return_counts=True)
  #menghitung indeks gini dari setiap partisi yang terbentuk
  gini_split = np.sum([(banyak[i]/np.sum(banyak))*hitung_gini(data.where(data[nama_fitur_split]==nilai[i]).dropna()[nama_fitur_kelas]) for i in range(len(nilai))])
  #mengembalikan nilai gini_split
  return gini_split

Ujilah fungsi **gini_split** menggunakan data_latih pada variabel **buying** dan variabel kelas bernama **class**.

In [12]:
#menghitung indeks gini dari berdasarkan fitur buying 
gini_split(data_latih, "buying", "class")

0.6536404135210501

## 4) Pembentukan pohon

Pembentukan pohon dilakukan secara rekursif. Seperti metode rekursif pada umumnya, perlu ditentukan kondisi berhenti terlebih dahulu. Kondisi berhenti pada pembentukan pohon adalah:


1.   Jika hanya ada satu kelas pada data, kembalikan kelas tersebut
2.   Jika fitur data  = 0 (tidak ada fitur yang tersisa), kembalikan kelas dari parent
3. Jika data kosong (tidak ada data), kembalikan kelas dengan frekuensi terbanyak

Selain kondisi berhenti tersebut, dilakukan pembentukan pohon secara rekursif menggunakan fungsi **buat_tree**.



In [13]:
#membuat fungsi untuk membuat tree secara rekursif
def buat_tree(data, data_awal, daftar_fitur, nama_fitur_kelas, kelas_parent_node = None):
  #jika hanya ada satu kelas pada data
  if len(np.unique(data[nama_fitur_kelas])) <= 1 :
    #mengembalikan kelas tersebut
    return np.unique(data[nama_fitur_kelas])[0]
  #jika data kosong
  elif len(data)==0:
    #mengembalikan kelas dengan frekuensi terbesar
    return np.unique(data_awal[nama_fitur_kelas])[np.argmax(np.unique(data_awal[nama_fitur_kelas],return_counts=True)[1])]
  #jika tidak ada fitur yang tersisa
  elif len(daftar_fitur)==0:
    #mengembalikan parent node
    return kelas_parent_node
  else :
    #mengambil kelas dengan frekuensi terbanyak sebagai parent node
    kelas_parent_node = np.unique(data[nama_fitur_kelas])[np.argmax(np.unique(data[nama_fitur_kelas], return_counts=True)[1])]
    #menghitung indeks gini dengan beberapa partisi
    nilai_split = [gini_split(data, fitur, nama_fitur_kelas) for fitur in daftar_fitur]
    #mengambil indeks fitur terbaik dengan nilai indeks gini terkecil
    index_fitur_terbaik = np.argmin(nilai_split)
    #mengambil fitur terbaik dari daftar fitur
    fitur_terbaik = daftar_fitur[index_fitur_terbaik]
    #membuat tree dengan parent fitur terbaik
    tree = {fitur_terbaik:{}}
    #membuat daftar fitur yang terdiri dari fitur yang ada tanpa fitur_terbaik
    daftar_fitur = [i for i in daftar_fitur if i != fitur_terbaik]
    for nilai in np.unique(data[fitur_terbaik]):
      #menghapus data ketika data berada di indeks yerbaik dan nilainya sama dengan nilai fitur terbaik
      sub_data = data.where(data[fitur_terbaik] == nilai).dropna()
      #membuat subtree dengan pemanggilan fungsi tree secara rekursif
      subtree = buat_tree(sub_data,data_awal,daftar_fitur,nama_fitur_kelas,kelas_parent_node)
      #menyimpan nilai subtree pada tree yang sudah terbentuk
      tree[fitur_terbaik][nilai]=subtree
  #mengembalikan tree yang sudah terbentuk
  return(tree)

Buatlah tree menggunakan data latih yang tersedia

In [14]:
#membuat tree dengan pemanggilan fungsi buat tree
tree = buat_tree(data_latih, data_latih, data_latih.columns[:-1],'class')

Tampilkan tree yang terbentuk. Gunakan library **pprint** untuk menampilkan dictionary secara teratur.

In [15]:
#melakukan pemanggilan library pretty print untuk mencetak tree
from pprint import pprint
#penampilkan tree dengan pretty print
pprint(tree)

{'safety': {'high': {'buying': {'high': {'maint': {'high': {'lug_boot': {'med': 'acc',
                                                                         'small': 'acc'}},
                                                   'low': {'lug_boot': {'big': 'acc',
                                                                        'small': 'acc'}},
                                                   'med': {'lug_boot': {'small': 'acc'}},
                                                   'vhigh': 'unacc'}},
                                'low': {'lug_boot': {'big': {'maint': {'high': 'vgood',
                                                                       'low': 'vgood',
                                                                       'med': 'vgood',
                                                                       'vhigh': 'vgood'}},
                                                     'med': {'maint': {'high': 'vgood',
                                            

## 5) Proses prediksi

Proses prediksi kelas pada data uji dilakukan dengan melakukan *tree traversal* sampai menemui leaf.

In [16]:
#melakukan prediksi dengan membuat fungsi prediksi yang berparameter datauji dan tree yang sudah dibuat
def prediksi(data_uji,tree):
  for key in list(data_uji.keys()):
    if key in list(tree.keys()):
      try:
        #prediksi data uji disimpan dalam variabel hasil
        hasil = tree[key][data_uji[key]]
      except:
        return 1
        #prediksi data uji disimpan dalam variabel hasil
      hasil = tree[key][data_uji[key]]
      #cek apakah hasil merupakan dictionary
      if isinstance(hasil,dict):
        #mengembalikan nilai prediksi
        return prediksi(data_uji,hasil)
      else:
        #mengembalikan nilai dari hasil prediksi
        return hasil

## 6) Proses Pengujian
Lakukan pengujian menggunakan data uji. Kelas pada data uji perlu dihapus dan data uji perlu diubah menjadi dictionary

In [17]:
#melakukan penghapusan kelas dan mengubah ke dictionary
data_uji_dict = data_uji.iloc[:,:-1].to_dict(orient = "records")

Lakukan pengujian terhadap keseluruhan data uji menggunakan looping.

In [18]:
#melakukan pengujian semua data uji menggunakan perulangan
hasil_prediksi_total = []
for i in range(len(data_uji_dict)):
  #menampung hasil prediksi pada variabel hasil_prediksi
  hasil_prediksi = prediksi(data_uji_dict[i],tree)
  #hasil prediksi ditambahkan pada list yang sudah dideklarasikan diawal
  hasil_prediksi_total.append(hasil_prediksi)

Bandingkan hasil prediksi dengan label sebenarnya. Hitunglah banyaknya data uji yang memiliki kelas prediksi sama dengan kelas sebenarnya

In [19]:
#menghitung prediksi yang tepat 
print("Total prediksi benar:",sum(hasil_prediksi_total==data_uji['class']))

Total prediksi benar: 44


In [20]:
#menghitung akurasi dengan membagi prediksi yang tepat dengan jumlah data yang diuji
akurasi = (sum(hasil_prediksi_total==data_uji['class']) / data_uji.shape[0])
#menampilkan jumlah data uji
print("Total data sebenarnya:", data_uji.shape[0])
#menampilkan nilai akurasi
print("Akurasi:", akurasi)

Total data sebenarnya: 52
Akurasi: 0.8461538461538461


## TUGAS
Pada tugas kali ini Anda diminta memodifikasi metode pembentukan tree yang telah Anda agar metode tersebut menggunakan information gain sebagai dasar percabangan. Lengkapilah kerangka source code di bawah ini

Lengkapi fungsi hitung_entropy

In [21]:
#membuat fungsi hitung entropi untuk menghitung nilai entrophy
def hitung_entropy(kolom_kelas):
  #menyimpan nilai kolom_kelas dalam variabel elemen dengan menghilangkan duplikat
  #menyimpan banyak data
  elemen, banyak = np.unique(kolom_kelas, return_counts=True)
  #mengitung nilai entrophy
  entropy =-1 * (np.sum([(banyak[i]/np.sum(banyak))*np.log2(banyak[i]/np.sum(banyak)) for i in range(len(elemen))]))
  #mengembalikan nilai entrophy
  return entropy

In [22]:
#pengecekan fungsi yang sudah dibuat berjalan dengan lancar
nilai=hitung_entropy(data_latih['buying'])
print(nilai)

1.8282851867679533


Lengkapi fungsi information_gain

In [23]:
#membuat fungsi hitung information gain untuk menghitung nilai information gain
def information_gain(data, nama_fitur_split, nama_fitur_kelas):
  #menyimpan nilai kolom_kelas dalam variabel elemen dengan menghilangkan duplikat
  #menyimpan banyak data
  nilai, banyak = np.unique(data[nama_fitur_split], return_counts=True)
  #menghitung nilai entrophy parent
  entrophy_parent = hitung_entropy(data[nama_fitur_kelas])
  #menghitung nilai entrophy partiki child yang terbentuk
  entrophy= np.sum([(banyak[i]/np.sum(banyak))*hitung_entropy(data.where(data[nama_fitur_split]==nilai[i]).dropna()[nama_fitur_kelas]) for i in range(len(nilai))])
  #menghitung nilai information gain
  information_gain = entrophy_parent - entrophy
  #mengembalikan nilai information_gain
  return information_gain 

In [24]:
#pengecekan fungsi yang sudah dibuat berjalan dengan lancar
ig=information_gain(data_latih,'safety', 'class')
print(ig)

0.5376036735052974


Lengkapi fungsi **buat_tree_ig**. Isinya sama persis dengan fungsi **buat_tree**, hanya saja penghitungan **gini_split** diganti dengan **information_gain**. Selain itu, percabangan dilakukan dengan menggunakan nilai **information_gain** **terbesar**

In [25]:
#membuat fungsi buat tree information gain
def buat_tree_ig(data,data_awal, daftar_fitur, nama_fitur_kelas,kelas_parent_node=None):
  #jika hanya ada satu kelas pada data
  if len(np.unique(data[nama_fitur_kelas])) <= 1 :
    #mengembalikan kelas tersebut
    return np.unique(data[nama_fitur_kelas])[0]
  #jika data kosong
  elif len(data)==0:
    #mengembalikan kelas dengan frekuensi terbesar
    return np.unique(data_awal[nama_fitur_kelas])[np.argmax(np.unique(data_awal[nama_fitur_kelas],return_counts=True)[1])]
  #jika tidak ada fitur yang tersisa
  elif len(daftar_fitur)==0:
    #mengembalikan parent node
    return kelas_parent_node
  else :
    #mengambil kelas dengan frekuensi terbanyak sebagai parent node
    kelas_parent_node = np.unique(data[nama_fitur_kelas])[np.argmax(np.unique(data[nama_fitur_kelas], return_counts=True)[1])]
    #menghitung information gain dengan beberapa partisi
    nilai_split = [information_gain(data, fitur, nama_fitur_kelas) for fitur in daftar_fitur]
    #mengambil indeks fitur terbaik dengan nilai information gain terbesar
    index_fitur_terbaik = np.argmax(nilai_split)
    #mengambil fitur terbaik dari daftar fitur
    fitur_terbaik = daftar_fitur[index_fitur_terbaik]
    #membuat tree dengan parent fitur terbaik
    tree = {fitur_terbaik:{}}
    #membuat daftar fitur yang terdiri dari fitur yang ada tanpa fitur_terbaik
    daftar_fitur = [i for i in daftar_fitur if i != fitur_terbaik]
    for nilai in np.unique(data[fitur_terbaik]):
      #menghapus data ketika data berada di indeks yerbaik dan nilainya sama dengan nilai fitur terbaik
      sub_data = data.where(data[fitur_terbaik] == nilai).dropna()
      #membuat subtree dengan pemanggilan fungsi tree secara rekursif
      subtree = buat_tree_ig(sub_data,data_awal,daftar_fitur,nama_fitur_kelas,kelas_parent_node)
      #menyimpan nilai subtree pada tree yang sudah terbentuk
      tree[fitur_terbaik][nilai]=subtree
  #mengembalikan tree yang sudah terbentuk
  return(tree)

Lakukan pembentukan tree menggunakan fungsi **buat_tree_ig**

In [26]:
#membuat tree information gain dengan pemanggilan fungsi yang sudah dibuat 
tree_ig = buat_tree_ig(data_latih,data_latih,data_latih.columns[:-1],'class')

Tampilkan tree yang terbentuk

In [27]:
#memncetaj tree yang sudah dibuat dengan pretty print format
pprint(tree_ig)

{'safety': {'high': {'buying': {'high': {'maint': {'high': {'lug_boot': {'med': 'acc',
                                                                         'small': 'acc'}},
                                                   'low': {'lug_boot': {'big': 'acc',
                                                                        'small': 'acc'}},
                                                   'med': {'lug_boot': {'small': 'acc'}},
                                                   'vhigh': 'unacc'}},
                                'low': {'maint': {'high': {'lug_boot': {'big': 'vgood',
                                                                        'med': 'vgood',
                                                                        'small': 'vgood'}},
                                                  'low': {'lug_boot': {'big': 'vgood',
                                                                       'med': 'vgood',
                                           

Lakukan pengujian menggunakan tree yang terbentuk

In [28]:
#melakukan pengujian semua data uji menggunakan perulangan
hasil_prediksi_total_ig = []
for i in range(len(data_uji_dict)):
  #menampung hasil prediksi pada variabel hasil_prediksi
  hasil_prediksi = prediksi(data_uji_dict[i],tree)
  #hasil prediksi ditambahkan pada list yang sudah dideklarasikan diawal
  hasil_prediksi_total_ig.append(hasil_prediksi)

In [29]:
#menampilkan nilai total jumlah yyang melakukan prediksi dengan benar
print("Total prediksi benar: ",sum(hasil_prediksi_total_ig==data_uji['class']))

Total prediksi benar:  44


In [30]:
#menghitung akurasi dengan membagi prediksi yang tepat dengan jumlah data yang diuji 
akurasi = (sum(hasil_prediksi_total_ig==data_uji['class']) / data_uji.shape[0])
#menampilkan jumlah data uji
print("Total data sebenarnya:", data_uji.shape[0])
#menampilkan nilai akurasi
print("Akurasi:", akurasi)

Total data sebenarnya: 52
Akurasi: 0.8461538461538461


# PERTANYAAN

Jawablah pertanyaan di bawah ini



1.   Amati tree yang dihasilkan dengan kriteria percabangan GINI dan Information Gain. Apa perbedaan tree yang dihasilkan dari kedua metode tersebut?
2.   Apakah penggunaan Information Gain dapat meningkatkan akurasi prediksi?



Jawaban :

1.   Tree yang dihasilkan oleh kedua metode memiliki perbedaan pada anak dari treenya, namun pada node root sama yakni kelas Safety yang menjadi node root. Perbedaan child dapat terlihat pada anak child safety yang medium memiliki child fitur buying kemudian memiliki anak fitur lug_boot dan maint pada tree gini. Sedangkan pada tree information gain anak child safety yang bernilai medium memiliki child fitur maint terlebih dahulu kemudian baru memiliki anak tree yang berisikan fitur buying dan lug_boot. Selain itu pada tree gini sub tree safety high buying low anak tree nya berupa fitur lug_boot dan anak tree lug_boot adalah maint sedangkan pada tree information gain sub tree safety high buying low anak tree nya berupa fitur maint  dan anak tree maint adalah lug_boot yang mana kebalikan dari tree Gini. Perbedaan percabangan ini disebabkan perbedaan pemilihan fitur dengan nilai Gini terendah dan pemilihan nilai information gain tertinggi yang berbeda. Metode information gain memiliki tingkat akurasi yang lebih baik daripada metode Gini, namun metode Gini memiliki tingkat presisi yang lebih baik daripada metode Information Gain.


2.   Dari percobaan yang telah dilakukan akurasi yang didapatkan sama, yakni dengan nilai prediksi benar sejumlah 44 data uji. Tingkat akurasi prediksi keduanya dapat terlihat perbedaan yang banyak ketika menggunakan dataset yang besar, pada dataset kecil seperti Car evaluation dataset yang hanya memiliki 260 baris data dan 5 kolom atau fitur dan kelas belum terlihat perbedaan pada hasil akurasinya . Berdasarkan beberapa sumber metode Information Gain memilki akurasi prediksi yang lebih baik daripada metode GINI. Namun, GINI memiliki tingkat presisi yang lebih baik daripada Information Gain. Sehingga keduanya dapat digunakan sesuai dengan kebutuhan masing-masing.
